app/components/contacts/AddNewContact.js

```diff
@@ -0,0 +1,222 @@
+  flex-direction: row;
+  justify-content: ${({ alignment }) => (alignment === 'right' ? 'flex-end' : 'flex-right')};
+`;
+
+type Field = {
+  title: string,
+  placeholder: string,
+  fieldName: string,
+  errorFieldName: string
+};
+
+type Props = {
+  modalMode?: boolean,
+  addToContacts: Action,
+  publicWalletAddress?: string,
+  resolve: ({ publicWalletAddress: string, nickname: string, email?: string }) => void
+};
+
+type State = {
+  publicWalletAddress: string,
+  nickname: string,
+  email?: string,
+  addressErrorMsg: ?string,
+  nicknameErrorMsg: ?string,
+  emailErrorMsg: ?string,
+  renderKey: number
+};
+
+class AddNewContact extends Component<Props, State> {
+  state = { ...initialState };
+
+  static getDerivedStateFromProps(props, prevState) {
+    if (props.publicWalletAddress && !prevState.publicWalletAddress) {
+      return { publicWalletAddress: props.publicWalletAddress };
+    }
+    return null;
+  }
+
+  render() {
+    const { modalMode } = this.props;
+    const { publicWalletAddress, nickname, addressErrorMsg, nicknameErrorMsg, emailErrorMsg, renderKey } = this.state;
+    const isSaveButtonDisabled = !publicWalletAddress || !nickname || !!addressErrorMsg || !!nicknameErrorMsg || !!emailErrorMsg;
+    return (
+      <Wrapper showBorder={!modalMode} key={`render_key_${renderKey}`}>
+        {!modalMode && (
+          <TitleWrapper>
+            <Title>Add New Contact</Title>
+          </TitleWrapper>
+        )}
+        {this.renderFields()}
+        <ButtonWrapper alignment={modalMode ? 'right' : 'left'}>
+          <SmButton text="Save to contacts" isDisabled={isSaveButtonDisabled} theme="orange" onPress={this.handleSave} style={{ width: 164 }} />
+        </ButtonWrapper>
+      </Wrapper>
+    );
+  }
+
+  renderFields = () => {
+    return <React.Fragment>{fields.map((field: Field) => this.renderSingleField(field))}</React.Fragment>;
+  };
+
+  renderSingleField = ({ title, placeholder, fieldName, errorFieldName }: Field) => {
+    const { publicWalletAddress } = this.props;
+    const { addressErrorMsg, nicknameErrorMsg, emailErrorMsg } = this.state;
+    const isPublicAddressFromProps = fieldName === 'publicWalletAddress' && !!publicWalletAddress;
+    const errorMessages = {
+      publicWalletAddress: addressErrorMsg,
+      nickname: nicknameErrorMsg,
+      email: emailErrorMsg
+    };
+    return (
+      <FieldWrapper key={fieldName}>
+        <TextWrapper>
+          <BoldText>{title}</BoldText>
+        </TextWrapper>
+        <SmInput
+          type="text"
+          isDisabled={isPublicAddressFromProps}
+          placeholder={placeholder}
+          defaultValue={isPublicAddressFromProps ? publicWalletAddress : ''}
+          errorMsg={errorMessages[fieldName]}
+          onChange={({ value }) => this.handleTyping({ value, fieldName, errorFieldName })}
+        />
+      </FieldWrapper>
+    );
+  };
+
+  getErrorMessage = ({ fieldName, value }: { fieldName: string, value: string }) => {
```

Comment:
this function does validation not just return error message or null
also think how to merge validations to single function without creating useless one time-use map
errorMessages object should be out of class const